### ISSUE TITLE
RFC: 3.0 API & Ideas

### ISSUE THREAD

**Issue by thebigredgeek** [visionmedia/debug#370]
Hey everyone,

I figured it's time to start discussing the API for V3.

A few fundamental ideas to get started:

* (No) Code written in ES2015 (minimal transpilation?, no es6 modules)
* (Yes) Robust test coverage
* (Yes) XO Code Styling
* (Yes) "Replaceable middleware" pattern for formatting, output, etc.  Extensible.   This allows hooks for stuff like logstash and client-to-server log streaming as desired.  Middleware is FIFO and debug comes with a default set for formatting, output, environment adaptation, etc.  Ex: 

```javascript
 //Stdout would be default console.log middleware
import debug, { StdoutMiddleware } from 'debug';
// Add some sort of new middleware element
log.use((data, next) => ... )
// replace Stdout (console.log) with console.info at same position in middleware array
log.replace(StdoutMiddleware, (data, next) => {
  console.info.apply(console, data);
  next(); 
});

```

* (Yes) "Child" debug instances (which inherit middleware from parent debug instances).  Ex: 
```javascript
const log1 = debug('foo');
log1.use((data, next) => ... );
const log2 = log.child('bar');
log2('hello world'); // => foo:bar hello world
```

* (Yes) Formal organized schema for CHANGELOG.md
* (Yes) Dead-simple generated documentation

* (Yes) Time locale configuration
* (Yes) Stdout moved off of APIs that are pending removal (https://github.com/visionmedia/debug/issues/280)
* (No) Ability to enable/disable debug instances
* (Maybe) API to overwrite/set environment variables by key (https://github.com/visionmedia/debug/issues/277)

Any other ideas are welcome.  I'd love to have a completed first pass draft by the end of next week.

**Comment by thebigredgeek** [comment]
https://github.com/visionmedia/debug/issues/343

**Comment by kilianc** [comment]
@thebigredgeek I am not sure if this came up somewhere else of there is already a solution in place but it would be nice to be able to switch between `STDERR` and `STDOUT` at the API level.

Example:

```js
import createDebug from 'debug'

const debug = createDebug('app:upload')

Promise.reject(new Error('Something terrible happened O_O'))
  .catch((err) => debug.error(err.message))

Promise.resolve({ status: 'ok' })
  .then((data) => debug(data.status))

```

**Comment by thebigredgeek** [comment]
@kilianc  Seems like that case could be solved via a combination of child instances and middleware:

```javascript
import debug, { StdoutMiddleware } from 'debug';

const appUploadLog = debug('app:upload');

const appUploadErrLog = appUploadLog
  .child('error')
  .replace(StdoutMiddleware, (data, next) => {
    console.error.apply(console, data);
    next()
  });
```

**Comment by kilianc** [comment]
@thebigredgeek that's fair. Not that far from what I do today:

```
const createDebug = require('debug')

module.exports = (label) => {
  const debug = createDebug(label)
  debug.error = createDebug(label)
  debug.error.log = console.error.bind(console)
  return debug
}
```

This seems a generic enough use case to consider adding a method to all debug instances by default.

For example, I like the `.child` method but it is not that different from:

```js
const debug = require('debug')
const NAME = process.env.DEBUG_NAMESPACE || 'app'

module.exports = (name) => debug(`${NAME}:${name}`)
```

It would be nice if could gather together the most common use cases and add these in the default implementation, even in the form of middleware for consistency. It's just cumbersome for me to carry around a setup file for debug (`require('../lib/debug')`) that always looks the same :D.

**Comment by thebigredgeek** [comment]
@kilianc yeah you could just create a "base" debug, implement middleware on that one, and then create child debug instances from that.  Solves the same problem and avoids mutable state within the library itself

**Comment by yamikuronue** [comment]
I think the most important thing to keep in mind is simplicity. This module is incredibly simple to use; you'll want to make sure it remains dead simple to use, and is as simple as possible to extend and configure.

**Comment by tj** [comment]
Beyond the basic pattern matching stuff I think it might be easier to just use an existing logger when it comes to contexts etc. I liked the module best before all the browser stuff when it was basically just a pattern-matched `console.log`. Anything beyond that I think almost encourages using it like a logger, which it's definitely not, it was originally just for debug-level stuff that would be a huge pain to propagate through your app to view what's going on.

**Comment by kilianc** [comment]
To add to @tj's point, that's the main reason why it defaults to `STDERR`, so it doesn't get in the way of your logs, correct?

The problem is that it looks pretty, it's dead simple to use and people dig it. Myself included, we have been misusing the library.

**Comment by tj** [comment]
@kilianc yep, it is (was at least) really just for quick sanity checks of like "why the hell is this stuck" etc haha.

**Comment by thebigredgeek** [comment]
@tj I agree for the most part.  I think the API should be dead simple.  Call a function with a string containing formatter escapes, etc.  Like the original lib was intended to work.  However, I think we could also add middleware and a few configuration options to satisfy the many people who use this lib as something beyond a simple stdout formatter, and I think we can do this without making the lib overly complex.  In fact, I think that would reduce complexity.  Currently the lib tries to take a bunch of weird edge cases into account and, IMO, do too much as a part of it's core functionality.  A lot of people love using debug as something more than a simple stdout wrapper, so I don't know if we should necessarily remove the ability to use it as something more so much as we should provide a dead simple API along with power user features that aren't required but are available and that scale well regardless of your project size... whether that is a reusable lib that uses debug in it's most basic form or a large project that uses power user features but also leverages the tiny surface area of the simplistic API.

**Comment by thebigredgeek** [comment]
@yamikuronue for sure.  Winston is a good example of a logging lib that has a way-too-big surface area for my taste.  I want to be able to pick up the lib and use it in under 2 lines of code, and I don't want to have to download more than 1 or 2 dependencies to go along with the lib.  At the same time, extensibility options for power users allow me to learn one logging lib both for simple and complex project cases.

**Comment by yamikuronue** [comment]
Maybe it would help to enumerate use cases and/or personas when you're looking at adding new features?

The thing that draws me to this module is that it's easy to scope your output to only get debug info for the bits you're working on actively, but you can still leave the debug statements in place permanently for when something you didn't expect to break does in a month. We use it in our pluggable software, so I might only care about a specific provider today, or a given plugin. Our users only want to see actionable output, from our main logger. But we might ask them to repeat a broken use case with debug on and send us the results.

**Comment by kilianc** [comment]
@yamikuronue use case is very coherent with the original idea of this library. Debug statements that you turn off in production. *A lot* of third party modules use `debug` to print debugging information that the user can turn on and off. For example `DEBUG=express:*`.

I think we should just discourage people from using `debug` for standard logging or create a new module.

**Comment by thebigredgeek** [comment]
@yamikuronue great idea!  The above is the persona I am most comfortable with.  In day to day work, I use debug both for libs and for large projects like REST APIs etc.  I've tried a lot of other logging libs like winston but they are really complex and I end up spend a good amount of time getting them set up properly which I shouldn't have to do.  At the same time, with debug, if I want to aggregate logs somewhere I have to use something like syslogd at the ops level which I really like to avoid if I can help it.  I also may need to format logs differently for ops people to be happy, etc.  BUT, I don't want to use a heavyweight logging lib that has a steep learning curve.

**Comment by thebigredgeek** [comment]
@kilianc honestly I am not sure if there is a point in discouraging people to use debug like that.  We already have so many conditional formatting clauses and adapters for different environments that implementing simple configuration kills multiple birds with one stone.  Unless we are talking about reducing debug down to a simple `console.log` wrapper and removing all of extra the functionality it already has, there is still a need to organize stuff properly even internally.  Might as well make something like that public so that we don't end up getting bombed with hundreds of feature requests and API creeping PRs every quarter

**Comment by kilianc** [comment]
@thebigredgeek same use case over here. Do you think it's worth it branch off to a different module and just refactor `debug` to make it more like a framework while keeping the original intent?

**Comment by thebigredgeek** [comment]
@kilianc honestly I don't know if that's a good idea.  We would either need to a) remove all of the functionality that exists and piss a bunch of people off or b) leave the code in it's current spaghetti form and continue merging in PRs that creep the scope of the API.  I think we can keep the current behavior while "passing the buck" with feature creep so that people can roll their own outputs, formatters, etc without having to hack debug or push through a PR for core functionality

**Comment by tj** [comment]
Formatters are kind of awkward since debug then gets globally modified magically, if people have formatting string issues it should go into Node core.

DEBUG_FD is out of scope too IMO, logs should pretty much always be stderr, certainly debug()'s ephemeral output. The other env vars seem useful.

**Comment by tj** [comment]
Looks like removing DEBUG_FD alone would scrap all of https://github.com/visionmedia/debug/blob/master/node.js#L153-L220

**Comment by yamikuronue** [comment]
I think it makes sense to go ahead and move the features that currently exist into a testable, extensible architecture. It probably makes sense at that point to look at what feature requests actually come in and see about what people would use. Docs also make perfect sense, and tests. I don't know if the child instance thing actually makes as much sense as it looks like, for example; the current functionality is doing just fine for many people.

**Comment by thebigredgeek** [comment]
@yamikuronue I am also on the fence about child instances.  I can see how it could be useful, and several people have asked for it in issues.  It isn't that painful to build, maintain, or test using closures and a bit of deferred recursion, and I don't really see the API for it ever mutating.  At the same time, it's something that will no doubt need to be maintained.

**Comment by thebigredgeek** [comment]
@TooTallNate would love to have your input on some of this

**Comment by medikoo** [comment]
> Code written in ES2015 with imports

Does it mean it'll be published to npm and transpiled to NCJS modules format (so it works well in Node.js)? It seems slightly controversial, taking into account that Node.js is main environment in which this project is used.

**Comment by thebigredgeek** [comment]
Most of ES2015 is now supported natively by node, including some 2016 features like async/await in node 7

**Comment by thebigredgeek** [comment]
But yes, it will be transpiled before publication

**Comment by medikoo** [comment]
> Most of ES2015 is now supported natively by node, including some 2016 features like async/await in node 7

Sorry, maybe I was not clear enough. I was referring strictly to _imports_ (ES modules), native support of that is not even on horizon in case of Node.js. At this stage it's not certain whether it'll happen at all.
It weird to use it for module dedicated mostly for Node.js

**Comment by medikoo** [comment]
Same way it'll be not be great to write it e.g. in CoffeeScript or TypeScript, as it's something low-level, for very broad audience.
It's best if it's written with API's that are known to work natively, without additional burden of transpilation.

**Comment by thebigredgeek** [comment]
It definitely adds some extra moving parts.  Imports definitely aren't the most important feature in ES2015/2016, but I am not sure if transpiling is really that big of a deal these days.  Most of the new stuff is being written in either transpiled JS or Typescript AFAIK.  Not saying we should necessarily follow that trend, but I wouldn't want to strictly avoid it on the basis of avoiding a transpiler for the sake of it, especially when the output when using something without a runtime is so close to the original source.  Not to even mention source-maps, though I think we will want to avoid anything that would require source-maps to debug.  Using a transpiler would also mean that we can safely build for multiple versions of Node, including backwards compatibility beyond 4, without having to worry about different feature sets within V8

**Comment by yamikuronue** [comment]
I am also strongly against transpiling. It makes everything more complicated; now the code you're running isn't the code you've written, making debugging harder, and the tooling is a PITA to configure.

Is it really that hard to just write code that's backwards compatible? I mean, I've been writing Node.JS code since 0.10; it's a pain in the butt to have to forgo the newer features, but it's not impossible, and the less complicated your libraries are, the better.

**Comment by thebigredgeek** [comment]
I think we can avoid transpiling for the most part.  We definitely don't need stuff like es6 modules.  But it would be nice to leverage arrow functions, constants, etc without worrying about what version of V8 node is running

**Comment by thebigredgeek** [comment]
Just a heads up, I am prefixing each line item above with yes/no/maybe to indicate draft state

**Comment by yamikuronue** [comment]
Bringing Babel into play gets you all the downsides, though, really. Do we have access to any stats about what versions of node people who depend on the module are using? We could say, if we're doing a major version number, that we're EOLing support for some older versions, like maybe anything older than Node 4, while we do. That'd leave the current version out there for backwards compatibility while allowing us to ditch babel. Node 4 gets you template strings, variable scoping, arrow functions, and promises.

**Comment by thebigredgeek** [comment]
I definitely think that's an option, but it's a tradeoff between process complexity and build complexity.

**Comment by medikoo** [comment]
I believe the sanest approach is to write this in plain JavaScript, that is natively understood by oldest engine we want support.
That would probably leave us with ES5 and some stuff from ES2015, which really is great enough, and I don't see how it can be a problem for any JavaScript'er.

**Comment by thebigredgeek** [comment]
Agreed.  Let's pass on transpilation for now.  Marking that as a `no`

**Comment by thebigredgeek** [comment]
So stuff to still discuss for v1 of draft:

* "Replaceable middleware" pattern for formatting, output, etc.  Extensible.   This allows hooks for stuff like logstash and client-to-server log streaming as desired.  Middleware is FIFO and debug comes with a default set for formatting, output, environment adaptation, etc.  Ex: 

```javascript
 //Stdout would be default console.log middleware
import debug, { StdoutMiddleware } from 'debug';
// Add some sort of new middleware element
log.use((data, next) => ... )
// replace Stdout (console.log) with console.info at same position in middleware array
log.replace(StdoutMiddleware, (data, next) => {
  console.info.apply(console, data);
  next(); 
});

```

* "Child" debug instances (which inherit middleware from parent debug instances).  Ex: 
```javascript
const log1 = debug('foo');
log1.use((data, next) => ... );
const log2 = log.child('bar');
log2('hello world'); // => foo:bar hello world
```

* Dead-simple generated documentation

* Time locale configuration
* Ability to enable/disable debug instances at runtime
* API to overwrite/set environment variables by key (https://github.com/visionmedia/debug/issues/277)

**Comment by yamikuronue** [comment]
Documentation: yes, yes yes. We're setting that up for all SockDrawer's stuff and it's a huge boon, especially once you go into a pluggable architecture.

Child debug instances: eh. I could go either way personally. Ditto on the API. I'm sure you have a reason for wanting them in the first place, but I don't see any of my projects using them.

Enable/disable at runtime: what's the use case here?

Time locale configuration: probably a good idea.

**Comment by thebigredgeek** [comment]
@yamikuronue yeah I also think the enable/disable thing is kinda weird.  I was asked for by several people but I think the same could be accomplished with pluggable architecture.

**Comment by thebigredgeek** [comment]
@yamikuronue what do you mean by "API"?

**Comment by tj** [comment]
re: enabling at runtime, I originally thought it would be cool to hit the proc with a signal or repl in to enable stuff at runtime, but I think that enters https://github.com/jstrace & https://github.com/tj/node-trace territory

**Comment by thebigredgeek** [comment]
Always sucks when that happens.  Architecture astronauts :(

**Comment by thebigredgeek** [comment]
Let's use (this)[https://github.com/visionmedia/debug/tree/v3] for v3

**Comment by thebigredgeek** [comment]
I personally think that child instances + middleware opens up quite a bit of flexibility when using this package for stuff beyond simple output.  I think that is worth adding for V3, especially since it shouldn't be a breaking change.

**Comment by thebigredgeek** [comment]
@tj to clarify on above, do you think enable/disable at runtime has merit?  It isn't a breaking change, so it's free other than maintenance work haha

**Comment by tj** [comment]
Probably not, jstrace with objects is definitely a more flexible option for debugging running stuff, especially with high-traffic APIs etc, then you can at least summarize data instead of just seeing thousands of log lines.

**Comment by yamikuronue** [comment]
@thebigredgeek  I meant "API to overwrite/set environment variables by key". Sorry to not be more clear :)

**Comment by thebigredgeek** [comment]
Alright everyone, I think we are mostly set on a "first draft" for a V3 API/design/style.  Here is a list stuff mapped to new issues to dive into details for each (if applicable).

* Robust test coverage
* Code style guide and ESLint for enforcement (https://github.com/visionmedia/debug/issues/397)
* Middleware pattern for organization of environment-specific output formatting etc, as well as end-user customization and extensibility (https://github.com/visionmedia/debug/issues/398)
* Child debug instances (https://github.com/visionmedia/debug/issues/399)
* Formal organized schema for CHANGELOG.md (https://github.com/visionmedia/debug/issues/400)
* Dead-simple generated documentation (https://github.com/visionmedia/debug/issues/401)
* Time locale configuration (https://github.com/visionmedia/debug/issues/402)
* Stdout moved off of APIs that are pending removal (https://github.com/visionmedia/debug/issues/280)

Thank you everyone for the help, and happy holidays!

**Comment by adoyle-h** [comment]
I think it will be a great feature to achieve the "Ability to enable/disable specific debugging namespaces at runtime".
Sometimes I want to debug a http server **step by step**. For examples,

```js
debug.enable('connect:*');
// check out the logs
debug.disable('connect:*');
debug.enable('http:body-parser');
// send request and check out the logs
debug.disable('http:body-parser');
debug.enable('http:response');
// send another request and check out the logs
// ...
```

Changing environment variables could also work. Stopping the server, setting different environment variables, restarting server again, and it maybe cost much time for waiting server to start. So it is complicated for the scenario. I desire to **debug without shutting down the process**.

`middleware` is a way to transform data in a pipeline. But how to extend the debug instance itself? To implement `debug.enable(namespace)` and `debug.disable(namespace)` methods. Could you provide any appropriate cases? @thebigredgeek

**Comment by tehsenaus** [comment]
A big +1 from me for "Ability to enable/disable specific debugging namespaces at runtime."

The ability to turn on fine-grained logging for specific components (including sub-dependencies) at runtime would be so great for when things are getting weird in production. Restarting the process isn't an option.

I think the memory leak concerns are overstated. Just cap the tracking array, and worst case is you have leaked a few function closures. It's negligible.

**Comment by thebigredgeek** [comment]
@tehsenaus wouldn't `enable`/`disable` methods attached to each debug instance provide what you want without the overhead of a tracking array and leak (however minor it might be)?  I'd prefer to keep debug internals as stateless as possible.

**Comment by bmiller59** [comment]
Feature requests:

* Debug levels (e.g. info, warn, trace) to display that can be configured on/off via the DEBUG environment variable

* Ability to add custom format timestamp by letting the user provide a function to provide the time stamp string. (Seems this could be achieved by proposed middleware/custom time locale)

I do not use debug as my logger, but on dev machines sometimes being able to specify custom levels of debugging is helpful. Timestamps for the same reason.

Thanks.

**Comment by freewil** [comment]
> Debug levels (e.g. info, warn, trace) to display that can be configured on/off via the DEBUG environment variable

I'd be interested in hearing others' thoughts on this, but one of the things I like about `debug` is the fact that it doesn't have debug levels to complicate things. While working on a project, I often run tests or the application with `DEBUG=myproject:*`, then sometimes when I really need to debug something I can then turn on the debug messages for a certain module that one of my module uses, to go deeper.

**Comment by thebigredgeek** [comment]
Seems to be like debug "levels" would be superfluous, given that the existing pattern matching method allows you to toggle on/off whatever you want.  Coupled with child debug instances, this accomplishes the exact same kind of functionality with the need for opinionated "levels"

**Comment by dougwilson** [comment]
Re levels, seems weird to have them at all in this module, mainly because typically "debug" is a log level, and since this module is called "debug" it would seem to me that stretching beyond that is a new logging module itself.

**Comment by bmiller59** [comment]
If something similar to debug levels can be done with child debug instances controlled via patterns in environment variable that would be great.

**Comment by simonhac** [comment]
trying not to lose track of desire for this to be a tiny and simple to use library:

+1 for log levels.

i use third party libraries that take a logger in their constructor, where the logger is expected to support `console.(log|info|error)` -- `.warn` would be nice too.

+1 for optional timestamps

+1 for optional filename

**Comment by emilio-martinez** [comment]
Would something like generating tree shakable code be something that would be possible? This would be particularly helpful for browser output. Given that `debug` works with env variables, could code become tree shakable when building for production?

**Comment by thebigredgeek** [comment]
@emilio-martinez tree shaking shouldn't really be necessary.  This is, and always will be, an incredibly tiny library.  Unless it is dead simple to implement, I think we should avoid engineering for something like that.

**Comment by thebigredgeek** [comment]
@simonhac agreed.  The lib will still be just as small and simple, if not more so as we won't need to include every little feature in the core lib as it will be extendable.  The API should remain the same for basic use, so the change shouldn't be strictly breaking other than the need to apply output middleware for special formatting cases etc.

**Comment by simonhac** [comment]
>@tj to clarify on above, do you think enable/disable at runtime has merit? It isn't a breaking change, so it's free other than maintenance work haha

+1 for enable/disable at runtime. i'm writing a plugin for a system that already specifies logging levels in config files. it's very unintuitive for the user to have debug settings in their `.conf` file and also have to setup environment variables.

(i'd be happy to help with this if help is needed.)

**Comment by tehsenaus** [comment]
Regarding enable/disable at runtime, if this functionality can be made available through middleware I'd be very happy to develop/support an external module for it.

I created "hot-debug" to enable this in the current version (via monkey patching the debug module):
https://github.com/tehsenaus/hot-debug

**Comment by simonhac** [comment]
@tehsenaus i just discovered your hot-debug a couple of days ago. i really like it -- thank you!
(i expect to send you a pull request today or tomorrow for a minor enhancement.)

**Comment by jhs** [comment]
Hi, all. Thanks so much for debug! TL;DR = Please keep the basic API as simple as ever; and please allow as many features as possible to be implemented in middleware. (My personal needs are server transports to file, and browser transports to HTTP, for example.)

I propose that the v3 effort embrace the fact that many successful software projects grow organically from simpler, more modest projects. Node.js itself and practically all the major packages follow this pattern, for example.

debug got successful for two reasons:

1. Just like console.log, with some bells, whistles: easy to understand; easy to ship
2. Same behavior, front-end and back-end

Most projects die. Some projects flourish. And so, it's only a partial answer to say, "This is not a logging framework. If you want logging, use that then." That answer punishes success: a project's adopting debug was the right choice at that time. It paid off. Now their code base is rife--infested!--with `debug()` calls and those help solve problems. If your code makes messages as it runs, and those messages help people solve problems--that's a logger. That's larval logging. But with the example project's success, the stakes and demands have risen and there are new challenges to solve.

Simple projects should use simple loggers. Advanced projects should use advanced loggers. And so debug v3 should answer the "tweener" projects, for which migrating to a full-blown framework, up and down the stack, sounds like a bad risk/reward trade-off (both developer time and code complexity). Those projects ask, "Can we please just make debug do a bit more?"

I would hope that the debug project answers this question by saying, "Yes, with any luck, you will outgrow me one day. But until then, you can get a great bang-for-buck with our middleware/advanced-features in v3."

**Comment by thebigredgeek** [comment]
https://github.com/visionmedia/debug/projects/1

**Comment by shakefu** [comment]
Thanks so much for Debug! We've been using it for a long while now and it's awesome.

I stumbled across this and just wanted to add my support for "child" debuggers, since we frequently  end up with a pattern like this, when we want that ability:

```javascript
const debug = require('debug')
// blah blah code
const dbg = debug('app')
// blah blah more code
this.dbg = debug('app:class')
```

We'd love to see something dead simple like the proposed `.child` or maybe even `.clone` or `.new` or something that doesn't imply hierarchy - just the ability to get a new `debug` instance without having separate module and instance references would cover our use cases.

Also glad to see keeping the library tiny and light weight is also a priority. Some of the ideas sound like treading into full on logger world, which is a use case better covered by bigger tools. I definitely don't want to have to think about whether it's a good idea to add `debug` to my dependencies, if it's going to bloat the project, etc.

Cheers, keep up the good work!

**Comment by tarwich** [comment]
**Global vs per-prefix ms diffs**
I see that #155 was closed in reference to this thread, but the conversation was not continued here. A lot of times I use debug to benchmark parts of my app and find out what part is killing performance. If this feature was available, I could have in-effect flame graphs for free as each child `debug()` would log time in its own namespace.

```js
// api.js
const debug = require('debug')('api.js');

export function loadSomething(id) {
  debug('Loading a something...');
  await someNetworkCallOrSomething(id);
  debug('...done'); // 50ms
}

// consumer.js
const debug = require('debug')('consumer.js');
const api = require('api.js');

loadAllSomethings() {
  debug('Loading all things...');

  for (let i = 0; i < 5; ++i) {
    api.loadSomething(i);
  }

  debug('...done'); // 250ms
}
```

I do hope it's something that's still on the radar.

**Comment by talkingtab** [comment]
I haven't seen a way to do this, maybe it already exists:
Lets say I have three high level components, app, db, auth -each defined with debug(server.auth) etc. What I want is set a "path" in each of these and enable only debugging for these. For example, i would like to enable debug(server.auth.logout), debug(server.db.logout) and debug(server.app.logout) and then specify in DEBUG="*.*.logout" and have only those statements visible. This would let me debug a pathway through my code, then turn on another path and debug that. 

THanks for debug. I like the simplicity of it and don't want to lose that, but it is a use case. Maybe for now just pipe the output through grep. :-)

**Comment by yamikuronue** [comment]
It takes wildcards; does `*logout` work?

**Comment by talkingtab** [comment]
I'll try that! Thanks!

**Comment by talkingtab** [comment]
What I don't understand is how I can have multiple pathways in the same 
file instrumented but only turn on one of them at a time.

So for example if I have a file auth, with a method "login" and another 
"logout" I would only want debugging on the "login" one.

I could I suppose have two debug statements:

/* auth.js */

dbglogout = require('debug')('dauth:api:logout')

dbglogin = require('debug')('dauth:api:login')

const login = ()=> {

     dbglogin("I'm logging in");

     db.foo();

}

const logut = () => {

     dbglogout("I'm logging out");

     db.bar();

}

DEBUG=*logout

Yes?

**Comment by ibc** [comment]
Hi, sorry if I missed something but, does `(No) Ability to enable/disable debug instances` means that such a feature is not even under consideration?

**Comment by thebigredgeek** [comment]
@ibc We are still considering `(maybe) API to overwrite/set environment variables by key`, which should do the exact same thing as `(No) Ability to enable/disable debug instances`.

**Comment by simonhac** [comment]
@thebigredgeek, following @ibc's comment, i am a contributor to a plugin for an application. the applications manages all config and hands the plugin a config object at initialisation time.

it is normal practice that users of the application turn logging on/off through the config object.

so i needed to be able to individually enable/disable debug 'channels' at runtime.

i extended @tehsenaus's excellent `hot-debug` (fork of `debug`) with this [PR](https://github.com/tehsenaus/hot-debug/pull/1) -- i'll fork it off into a project one day soon.

basically, i'm hoping that v3 will allow (either directly or via its own plugin middleware layer):

```javascript
var debug = require('debug')('myPlugin')

...

function myInit(config) {
  debug.enable(config.debugEnabled)
}
```

**Comment by VBAssassin** [comment]
**Title:** Optimised debug information collection

**Use Case:** Sometimes producing debug output is extra intensive on resource. If it won't be displayed (debug environment var not set or doesn't match) there is no point in generating the debug info. Such heavy debug information could be binding prepared queries params to produce the actual query for debug, etc. For this reason, having the ability to wrap inside a callback would enhance any such performance issue by only executing debug info when it will be displayed.

**Example:**

`// Normal behaviour.
debug("Hello World"); 

// Only if this should be displayed will it execute the callback
debug(function() {
       return mysql.format(sql,values);
});`

**Comment by stefcameron** [comment]
What @VBAssassin wrote is a very good idea. I've used this to great effect in a logging class I wrote for use in many projects I've worked on. It also helps mitigate the overhead in the case where logging is disabled, or that particular logger is disabled.

**Comment by jfseb** [comment]
Apologizes or enlighten me if this is already in: 

I often face the issue that debug instrumentation bogs down performance.
```
var debuglog = require('debug')('myfile') 
debuglog(" inspect this > " + JSON.stringify(bitstuff, undefined,2)); // ooh, hefty Stringify
```

So i end up with 
`debuglog( debuglog.enabled ?  " inspect this > " + JSON.stringify(bitstuff, undefined,2) : "") ;` 
or similar. 

Suggestion is to change the signature to evaluate (lazily) if passed a function 

```
debuglog( () => `Inspect this > " + JSON.stringify(bitstuff, undefined,2) ) ;
``` 

Currently it recognizes the arg as a function and outputs: 
  myfile [Function] +0ms
which is not that much use. So either be incompatible or add a wrapper evalfunc

```
var debuglog = require('debug')('myfile').evalfunc;
debuglog( () => `Inspect this > " + JSON.stringify(bitstuff, undefined,2) ) ;
```

**Comment by ibc** [comment]
>```
>var debuglog = require('debug')('myfile').evalfunc;
>```

Couldn't it just internally check `typeof arguments[0] === 'function'`?

**Comment by TooTallNate** [comment]
@jfseb Use the `%j` or `%O` formatters:

```js
debuglog(" inspect this > %j", bitstuff) ;
```

That way you're just passing in an argument to a noop function when debug is not enabled. Then the processing happens under the hood when debug is enabled.

**Comment by PixnBits** [comment]
@TooTallNate works for this simple JSON case, but not sure about @VBAssassin's [example](https://github.com/visionmedia/debug/issues/370#issuecomment-294376790)

**Comment by jfseb** [comment]
My recommendation was actually a duplicate, thanks @PixnBits and @VBAssassin.

**Comment by tehsenaus** [comment]
@VBAssassin Can you not already do this?

```
if ( debug.enabled ) {
    debug(mysql.format(sql,values));
}
```

IMO the above is more readable, and more efficient as you're not creating a closure.

**Comment by StreetStrider** [comment]
@tehsenaus your code is lack encapsulation, it requires to know about `enabled` and also write an `if` every time; while @VBAssassin follows the common idea behind `debug` that «debug works only in debug» (in that case a function is passed) and pretty encapsulated.

**Comment by talkingtab** [comment]
I like the simple syntax of the closure- I'm much more likely to use it. 
And because it uses the same form as console.log() I don't have to 
remember a special syntax.

**Comment by tehsenaus** [comment]
@talkingtab how is it the same form as console.log?

@StreetStrider can you comment on readability? As far as encapsulation goes, I think ```.enabled``` wouldn't be a surprising or controversial addition to the public API of any logging library.

**Comment by StreetStrider** [comment]
@tehsenaus, to be honest, I think closure got better readability too. But it depends on person itself (partly). 🤔  It's like `for`-loops against `map`/`forEach`. Some people found `for`-loop more readable.

**Comment by talkingtab** [comment]
I just mean they take the same arguments:

**Comment by stefcameron** [comment]
@StreetStrider @tehsenaus Another way of writing this, given `debug.enabled`, would be:
```javascript
debug.enabled && debug(mysql.format(sql, values);
```
That might be a little more friendly, visually, and it still ensures that whatever code is on the right side of `&&` is only executed if debug is enabled. So you don't need a closure in this case. This form would also enable code optimizations with transpilers that support tree shaking. Telling the transpiler that `debug.enabled === false` would allow it to stip-out all the debug statements from the resulting code.

**Comment by 131** [comment]
debug simplicity come from the initial, user defined DEBUG env var. "What i care about".
I suggest to use a simple process.env.DEBUG_FORMAT (from tj/progress inspiration) that might default to
```
process.env.DEBUG_FORMAT = ":timestamp :namespace :body :timediff";

having additionnal formaters like
::namespace_colored
:timestamp_iso
```
Registering additional formaters should be part of debug API, yet, leaving the user choosing, very simply, what he desired (the same way he choose the namespaces he cares about)

**Comment by fab1an** [comment]
Hi, i have come to this thread, because I noticed that time-stamps were off in my logs.

Speaking from many years of experience, I wouldn't use any built-in Date primitives. 
The most sensible solution for doing anything date-related is using the java8-time API. There is a port for javascript https://js-joda.github.io/js-joda
(Yes, I've used `moment` in the past)

To prove my point, here is how js-joda does everything right without any configuration:
```
const {ZonedDateTime} = require("js-joda").use(require("js-joda-timezone"))
const debug = require("debug")("")
debug.useColors = false
debug("<--- NODE | JS-JODA --->   ", ZonedDateTime.now().toString())
```

Output:
```
Sat, 05 Aug 2017 06:48:28 GMT  <-- NODE | JS-JODA --> 2017-08-05T08:48:28.541+02:00[Europe/Vienna]
```

js-joda got the correct timezone for my system, which is everything I want for a logging-library.

**Comment by Qix-** [comment]
Just throwing this out there - since this will be a major release I assume, the LTS schedule for Node should be closely followed. Right now, Node 5 is in LTS, which means es2015 should be safely distributable without the need for transpilation.

**Comment by yamikuronue** [comment]
@Qix- You mean Node 4? 

Also: this is transpiled for the frontend, as well as being available on the backend. It's safest and least problematic to stick to a widely-supported subset of JS when making code available for three platforms (Node, Electron, and front-end).

**Comment by thebigredgeek** [comment]
@Qix- haha yeah I wish it was that simple.  This lib is used in electron, browser, etc as well as Node.  So probably not going to work out so well without transpilation :(.  That said, because of the number of platforms we are supporting I do think there could be value in using a compiler step to normalize interfaces between environments.  @TooTallNate

**Comment by yamikuronue** [comment]
@thebigredgeek I mean, I did remove the transpilation step by just rewriting the thing in ES5 code ;P

**Comment by Qix-** [comment]
@yamikurone Node 4 is in maintenance LTS. Node 5 is currently the active LTS version.

**Comment by yamikuronue** [comment]
@Qix meaning Node 4 still has support. My company is skipping 5 because 4 is still in support and we might as well go to 6 as we upgrade in anticipation of losing that support. So I'm hesitant to recommend the idea that 5 is safe to assume for a widely used utility.

**Comment by Qix-** [comment]
Oops you're right @yamikuronue - 5.x is skipped, 6.x is now active LTS. 4.x is still maintenance, which means versions of software that still work for 4.x should backport any major bugfixes to it.

**Comment by paulrobello** [comment]
I think it would nice to be able specify my own color for a namespace or even a single call to log some output
